New interface using mostly Python code

mducle opened the pull request:
New Matlab interface which uses a Matlab wrapper over mostly Python code.

The Python code is in `brillem.py` which contains two functions `create_bz` and `create_grid` which makes constructing a `BrillouinZone` and `BZ[Trellis,Nest,Mest]Q[dd,dc,cc]` objects easier.

The envisage syntax is:

```matlab
bz = brille.create_bz(a,b,c,alpha,beta,gamma,spg)
grd = brille.create_grid(bz, 'node_volume_fraction', 1e-5)
plot(grd)
```

Help commands will redirect to Python docstring and the `pydoc` system.

----

Interface is now ready. Legacy code is still in the `+brillem` folder but new version works with `brille>=0.5.2` (which has the `.utils` functions) and [`mducle/spinw@b0c3e86`](https://github.com/mducle/spinw/tree/b0c3e8670a1513d8655605fa3f80ea182cfebcb8).

`+light_python_wrapper` is included as a submodule.
`requirements.txt` contains the required brille minimum version.

There is a script, `release.m` which will create a Matlab toolbox, `mltbx` file, and then create a github release and upload this toolbox. We can then link the github release page to the Matlab file exchange and users can then search for and install brillem as an "Add-on". The process for creating a release is:

```matlab
!git tag v0.2.0
setenv('GITHUB_TOKEN', '<PERSONAL_ACCESS_TOKEN>')
release('upload')
```

----

What's now needed:

* Push `brille` 0.5.2 to PyPI (latest version there is 0.5.0 which doesn't have the `.utils`).
* Merge this PR and create a release
* Link the github releases to the File Exchange
* Merge spinw changes to mainline spinw-3 and create a new release.

mducle committed:
New interface using mostly Python code

mducle committed:
Add SpinW support

Add "override" facility to light_python_wrapper
  used to overload python "fill" method with Matlab one
  to ensure correct type / dimensions of input
Add checks/conv for expect int/long from func signature
Add m2p; p2m now unwraps light_wrapper correctly
Need to use with mducle/spinw@25825a48f8

mducle committed:
Update grid to auto transpose interpolate inputs

mducle committed:
Bugfix: interpolate() options; grid type logic

mducle committed:
Fix: bz can use row or column vec; Add test scripts

mducle committed:
Update to new brille python

`brillem.py` moved to `brille.utils`.

mducle committed:
Create toolbox

Change to using `light_python_wrapper` submodule
    Removed `light_python_wrapper` code in +brille
Add code to install `brille` Python
Add code to create a mltbx toolbox
Add Contents.m help file to +brille
Add checks to see if `brille` Python installed
Add CHANGELOG.md
Brille version now stored in `requirements.txt`
Add `release.m` script to create a github release

g5t commented on +brille/private/required_modules.m:
```diff
@@ -0,0 +1,3 @@
+function out = required_modules()
+    out = {'brille', 'TO_BE_DETERMINED'};
```
Will leaving `'TO_BE_DETERMINED'` here throw an error in `brille_on` or `install_python_modules`?

mducle replied:
Ah... I should have noted that `TO_BE_DETERMINED` and `TO_BE_REPLACED` are both substituted in `release.m` for the actual package. If you're running with a "development" version (e.g. in the main repo directory) the `TO_BE_DETERMINED` will cause an error. I'll just rename it `0.0.0`.

g5t replied:
`'TO_BE_DETERMINED'` only appears in `required_modules.m` if you run, e.g,

``` sh
find . -type f -exec grep -H "TO_BE" {} \;
```

you get the output:
```
./+brille/private/required_modules.m:    out = {'brille', 'TO_BE_DETERMINED'};
./mltbx/brillem.prj:    <param.version>TO_BE_REPLACED</param.version>
./release.m:    text = regexprep(text, 'TO_BE_REPLACED', version);
```

And I see now that you replace the whole contents of `required_modules.m` when running `release`, so I guess indicating only the major version, `0`, that the (unreleased) package is intended to work with might be sufficient since `verify_python_modules` *should* expand `0` to the equivalent of `0.0.0`.

mducle replied:
yeah, it was such a short file it was easier just to rewrite it... but I'll change it so it does a proper string replacement in case we want to change the contents of `required_modules.m` later.

g5t commented on mltbx/brillem.prj:
```diff
@@ -0,0 +1,87 @@
+    <param.email>duc.le@stfc.ac.uk</param.email>
+    <param.summary>Interface to the Brille library for symmetry operations and Brillouin zone interpolation.</param.summary>
+    <param.description>Interface to a C++/Python code (Brille) for symmetry operations and Brillouin zone interpolation.</param.description>
+    <param.version>TO_BE_REPLACED</param.version>
```
Does `TO_BE_REPLACED` get automatically replaced by the package version, or is this here to be edited before the first real release?
If the latter, will the Mathworks File Exchange accept an upload with an invalid version number? If it will, is there a way to prevent accidentally publishing this there without a valid `param.version`?

g5t replied:
I see now that this gets automatically updated with the version number by `release.m` 🤦

mducle replied:
No worries...

g5t commented on release.m:
```diff
@@ -0,0 +1,178 @@
+    text = fread(fid, inf, '*char')';
+    fclose(fid);
+    fid = fopen('brillem.prj', 'w');
+    text = regexprep(text, 'TO_BE_REPLACED', version);
```
Ok, I should have looked a bit further before asking about 'TO_BE_REPLACED'.

g5t reviewed (approved):
With my concerns over `'TO_BE_DETERMINED'` and `'TO_BE_REPLACED'` managed I think this all looks good.

It might be nice to automate running your tests, but getting a GitHub runner up and running with MATLAB and spinW might be challenging. Though maybe [not impossible](https://github.com/matlab-actions/overview).

mducle commented:
Yeah, it should be ok on Linux. I'd have to write some tests that can actually fail first! (The current "tests" just run through some examples and don't check outputs).

mducle committed:
Review fixes

* Fix required_modules substitutions
* Made the release created by release.m a _draft_
* Also add function in release.m to restore
  required_modules.m and brillem.prj to their original
  state so the variable substitions always work.
